### ISSUE TITLE
Moving diagrams too slow due to collision detection

### ISSUE THREAD

**Issue by ceraolo** [OpenModelica/OpenModelica#11887]
### Description
Moving diagrams is too slow.

Consider Powergrids.Examples.IEEE14bus.IEEE14busStaticNetwork

On my PC moving it (after having done a CTRL-a to select all) with an arrokey requires 57 or 72s depending on whether I use 1.02 or the 2.0.0-develop version.

By comparison, this is nearly instantaneous on dymola.

I suppose this difference is due to the fact that when we move a diagram OMEdit asks info to OM for models and submodels, a thing that is not necessary.

If my explanation is right, enhancing motion operations by orders of magnitude should be very easy...

Times measured using  OM 1.24-dev-214

**Comment by adeas31** [comment]
I just did a quick test and it seems that `setElementAnnotation` and `updateConnectionAnnotation` take way too much time. @perost can you please check if we can improve this somehow?

**Comment by adeas31** [comment]
Bit more debugging shows that the real bottleneck is collision detection. I will see how we can improve that.

**Comment by perost** [comment]
Another issue is that e.g. `setElementAnnotation` and `updateConnectionAnnotation` updates the Absyn and invalidates the SCode. But `Interactive.evaluateExpr` builds a new environment from the SCode for each API call, which causes the whole loaded Absyn to be translated to SCode for every API call in this case.

Executing the annotation API calls takes about 10s (but having more libraries loaded will make this slower), disabling the invalidation of the SCode reduces it to 0.01s.

It's not clear to me if we really need all the loaded models in the environment we build for evaluating API calls, maybe @adrpo knows better? If we're only using the environment to look up API functions and variables we could maybe just cache it instead of recreating it every time.

**Comment by casella** [comment]
> Another issue is that e.g. `setElementAnnotation` and `updateConnectionAnnotation` updates the Absyn and invalidates the SCode. But `Interactive.evaluateExpr` builds a new environment from the SCode for each API call, which causes the whole loaded Absyn to be translated to SCode for every API call in this case.

This is of course something we should avoid.

> Executing the annotation API calls takes about 10s (but having more libraries loaded will make this slower), disabling the invalidation of the SCode reduces it to 0.01s.

That's more like it 😃 

> It's not clear to me if we really need all the loaded models in the environment we build for evaluating API calls, maybe @adrpo knows better? If we're only using the environment to look up API functions and variables we could maybe just cache it instead of recreating it every time.

This is a bit too technical for me, I'm afraid. However, there is for sure one point here: if you select a whole diagram and, say, move it up by 2 units, this operation must be carried out for each and every component, but it is an atomic operation. Even if invalidating the SCode (whatever that entails) is necessary, it should be done only once at the end of this atomic operation.

**Comment by ceraolo** [comment]
Other simple operations on IEEE14busStaticNetwork (PowerGrids 2.0.0-develop) that should in principle take a split second e.g. 0.1s) and instead get all around 10 seconds (measured on the same PC used tor measuring the times shown in the ticket's description):

- delete bus1- L1to5 a connection   (and the subsequent CTRL-Z)
- change GEN1 SNom (and the subsequent CTRL-Z)
- delete Load12  (and the subsequent CTRL-Z)

**Comment by casella** [comment]
@perost is trying to find a way to fix this

**Comment by adeas31** [comment]
PR #11901 fixes the issue of collision detection. The response time goes down from 5-6 mins to few secs. We can further improve it with API changes that @perost suggested.

**Comment by casella** [comment]
Great!

@adeas31 please cherry-pick #11901 to maintenance/v1.22.

Follow-up ticket for API issue: #11902.

**Comment by adeas31** [comment]
Ported to maintenance/v1.22

**Comment by perost** [comment]
#11906 improves the performance of the API, and together with the previous fix by @adeas31 moving is now pretty responsive. It's not perfect though, spamming the arrow keys seems to make it bog down quite a lot and gives warnings that it couldn't preserve the formatting of the model for some reason. But it's at least usable now.

**Comment by ceraolo** [comment]
I've just made checks (some insomnia helped).
Thank you, @adeas31  and @perost: now moving submodels and entire diagrams have become much faster! I would say around two orders of magnitude.

As @perost noted, there is still the issue related to pretty-printing algorithm (rather annoying).
Moreover, operations mentioned in  https://github.com/OpenModelica/OpenModelica/issues/11887#issuecomment-1913657986 have improved only a little (they take around 6 s now on may PC, which is still too much, IMO)

Tomorrow I plan to open two specific tickets regarding these remaining issues. I suppose this one (and #11902) can now be closed.
